### ISSUE TITLE
Message.properties() not working as expected in Python.

### ISSUE THREAD

**Issue by lelabo-marc** [apache/pulsar#3524]
### **Describe the bug**

TL;DR:

1. 

`<class '_pulsar.map_indexing_suite_MessageStringMap_entry'>.data()`
 is rising the following exception:

```
Traceback (most recent call last):
  File "/Applications/PyCharm.app/Contents/helpers/pydev/_pydevd_bundle/pydevd_exec2.py", line 3, in Exec
    exec(exp, global_vars, local_vars)
  File "<input>", line 1, in <module>
Boost.Python.ArgumentError: Python argument types in
    None.data()
did not match C++ signature:
    data(std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > {lvalue})
```

2. 

If you want message properties as a dictionary, use the following line of code:
`properties = {property.key(): msg.properties()[property.key()] for property in msg.properties()}`

### **The problem details**

The method description in the documentation:

> Return the properties attached to the message. Properties are application-defined key/value pairs that will be attached to the message.

In python, key/value implicitly mean that the method should return a dictionary or dictionary-like object, right ?

**The Message.properties() does not.** It return an object which is under the hood generated by boost in the [cpp-python interface](https://github.com/apache/pulsar/blob/master/pulsar-client-cpp/python/src/message.cc):

`    class_<Message::StringMap>("MessageStringMap").def(map_indexing_suite<Message::StringMap>());`

It is pretty hard to find a good documentation on what is actually generated in python and the object is hard to inspect correctly.

I will describe what is actually available and what should be expected:

### **Setup**
Producer:
```
import pulsar

client = pulsar.Client('pulsar://localhost:6650')
producer = client.create_producer('my-topic')
producer.send(('Hello you-%d' % i).encode('utf-8'), properties=dict(key1="value1", key2="value2"))
client.close()
```
Consumer:
```
import pulsar

client = pulsar.Client('pulsar://localhost:6650')
consumer = client.subscribe('my-topic', 'my-subscription')
msg = consumer.receive()
```

### **Lets dive in**

Ok, from the [python api doc](https://pulsar.apache.org/api/python/), the message properties sent with the producer should be accessible with the following code:
```
>>> msg.properties()
<_pulsar.MessageStringMap object at 0x1057435d0>
```

So the method return a unknown object from the cpp library.
It is not a dict, so lets try to cast it...

```
>>> dict(msg.properties())
Traceback (most recent call last):
  File "/Applications/PyCharm.app/Contents/helpers/pydev/_pydevd_bundle/pydevd_exec2.py", line 3, in Exec
    exec(exp, global_vars, local_vars)
  File "<input>", line 1, in <module>
TypeError: cannot convert dictionary update sequence element #0 to a sequence
```

This object does not support dict().
However it does support list() which allows us to start accessing the content of this std::map.

```
>>> list(msg.properties())
[(key1, value1), (key2, value2)]

>>> list(msg.properties())[0]
(key1, value1)
```

The objects contained in std::map are std::pairs, translated by boost we have an object that looks like a tuple but it can not be used directly with index or by casting to a real list/tuple.
```
>>> type(list(msg.properties())[0])
<class '_pulsar.map_indexing_suite_MessageStringMap_entry'>

>>> list(msg.properties())[0][0]
Traceback (most recent call last):
  File "/Applications/PyCharm.app/Contents/helpers/pydev/_pydevd_bundle/pydevd_exec2.py", line 3, in Exec
    exec(exp, global_vars, local_vars)
  File "<input>", line 1, in <module>
TypeError: 'map_indexing_suite_MessageStringMap_entry' object does not support indexing

>>> list(list(msg.properties())[0])
Traceback (most recent call last):
  File "/Applications/PyCharm.app/Contents/helpers/pydev/_pydevd_bundle/pydevd_exec2.py", line 3, in Exec
    exec(exp, global_vars, local_vars)
  File "<input>", line 1, in <module>
TypeError: 'map_indexing_suite_MessageStringMap_entry' object is not iterable
```

By inspecting the object with an IDE, two methods seems available: key() and data().
The first one is working as expected but the second one is not. It is raising an error because some mapping from C++ to Python is missing.
```
>>> list(msg.properties())[0].key()
'key1'

>>> list(msg.properties())[0].data()
Traceback (most recent call last):
  File "/Applications/PyCharm.app/Contents/helpers/pydev/_pydevd_bundle/pydevd_exec2.py", line 3, in Exec
    exec(exp, global_vars, local_vars)
  File "<input>", line 1, in <module>
TypeError: No Python class registered for C++ class std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >
```

This is the main problem with this method. Ok it may not behave like a dictionary but it should work as expected when using the available methods.

The only way available to access the value of the properties is:

`msg.properties()['key1']`

Thus the given methods to be able to extract keys, values and build a python dict.

```
>>> keys = [property.key() for property in msg.properties()]
['key1', 'key2']

>>> values = [msg.properties()[key] for key in keys]
['value1', 'value2']

>>> properties = {property.key(): msg.properties()[property.key()] for property in msg.properties()}
{'key1': 'value1', 'key2': 'value2'}
```

###**Expected behavior / Possible fixes**

- [ ] More information in the doc on which type of object is returned and how to use it.
- [ ] Fix the data() method and the mapping issues.
- [ ] Make Message.properties() return a python dict, created for example using the method:
`properties = {property.key(): msg.properties()[property.key()] for property in msg.properties()}`

### **Desktop**
 - OS: Mac OS

### **Additional context**
- I tried the python client available with pip (2.2.1).
- I tried to compile and install it from the github/master branch.

**Comment by merlimat** [comment]
Thanks for reporting. Will try to get a fix for this in 2.3
